### PR TITLE
[PGO] Add option to always instrumenting loop entries

### PR DESCRIPTION
**Pull request by ronryvchin**
This patch extends the PGO infrastructure with an option to prefer the instrumentation of loop entry blocks.
This option is a generalization of https://github.com/llvm/llvm-project/commit/19fb5b467bb97f95eace1f3637d2d1041cebd3ce,
and helps to cover cases where the loop exit is never executed.
An example where this can occur are event handling loops.

Note that change does NOT change the default behavior.

### CODE REVIEW

File: llvm/include/llvm/Transforms/Instrumentation/CFGMST.h
Comment (Dinistro) [inline llvm/include/llvm/Transforms/Instrumentation/CFGMST.h:162]: Nit: Maybe explain in a small comment what is happening here.

**Review by Dinistro — COMMENTED** [review]
Dropped a bunch of nit comments. The general idea makes sense to me and I think we can soonish open this for general reviews of people experienced with these components.

**Review by Dinistro — COMMENTED** [review]
Nice work 🙂 
Dropped one more nit comment. Apart from this, this should be ready for a review of owners/maintainers of the PGO infra.

File: llvm/lib/Transforms/Instrumentation/PGOInstrumentation.cpp
Comment (ellishg) [inline llvm/lib/Transforms/Instrumentation/PGOInstrumentation.cpp:2000]: Honestly I'm not sure why we don't simply pass the `FAM` rather than these lambdas, but this probably shouldn't block your work.

File: llvm/include/llvm/Transforms/Instrumentation/CFGMST.h
Comment (ellishg) [inline llvm/include/llvm/Transforms/Instrumentation/CFGMST.h:314]: I wonder if it makes sense to add an assert here that each loop entry will be instrumented.

File: llvm/test/Transforms/PGOProfile/loop3.ll
Comment (ellishg) [inline llvm/test/Transforms/PGOProfile/loop3.ll:1]: Can you also add a test that we can consume these profiles with loop entries?

**Review by ellishg — COMMENTED** [review]
I think we also need to add a new variant mask for instrumenting loop entries. Otherwise we won't be able to match the profiles they generate correctly

https://github.com/llvm/llvm-project/blob/dd78d7c7be5b8948cf5841e8033e59adebf230ad/llvm/lib/Transforms/Instrumentation/PGOInstrumentation.cpp#L439-L441

File: llvm/include/llvm/Transforms/Instrumentation/CFGMST.h
Comment (Dinistro) [reply llvm/include/llvm/Transforms/Instrumentation/CFGMST.h:314]: Could you elaborate what exactly you mean by this? Assert in the end that all loop entries are instrumented or what exactly?

File: llvm/include/llvm/ProfileData/InstrProfReader.h
Comment (Dinistro) [inline llvm/include/llvm/ProfileData/InstrProfReader.h:126]: I think it would be sensible to add a comment to this function, even when it's somewhat trivial.

File: llvm/include/llvm/Transforms/Instrumentation/CFGMST.h
Comment (ellishg) [reply llvm/include/llvm/Transforms/Instrumentation/CFGMST.h:314]: If `InstrumentLoopEntries` is true, then each loop entry should be instrumented. I'm asking if we should add a compile time assert that indeed each loop entry will be instrumented after we compute the minimum spanning tree.

File: compiler-rt/include/profile/InstrProfData.inc
Comment (ellishg) [inline compiler-rt/include/profile/InstrProfData.inc:749]: To maintain backwards compatibility these bits cannot be changed. I guess we need to expand the mask to support more bits.

File: llvm/test/Transforms/PGOProfile/Inputs/thinlto_indirect_call_promotion.profraw
Comment (ellishg) [inline llvm/test/Transforms/PGOProfile/Inputs/thinlto_indirect_call_promotion.profraw:1]: This seems to be unused.

File: llvm/test/Transforms/PGOProfile/loop3.ll
Comment (ellishg) [inline llvm/test/Transforms/PGOProfile/loop3.ll:2]: Instead of creating new files in the `Inputs` directory, lets use `split-file` to inline the `.proftext` files into this test.

https://llvm.org/docs/TestingGuide.html#extra-files

Comment (ellishg) [reply llvm/test/Transforms/PGOProfile/loop3.ll:2]: Also, it might be cleaner to split this into two separate tests: `PGOProfile/loop-entry-gen.ll` and `PGOProfile/loop-entry-use.ll`.

File: llvm/test/Transforms/PGOProfile/loop3.ll
Comment (ellishg) [inline llvm/test/Transforms/PGOProfile/loop3.ll:32]: You can add the `--implict-check-not="@llvm.instrprof.increment"` to your `FileCheck` command to reduce the number of check lines needed.

**Comment by ellishg** [comment]
> @ellishg Do you know how `llvm-project/llvm//test/tools/llvm-profdata/Inputs/function-entry-coverage.profdata` was generated? The new bit broke `llvm-project/llvm/test/tools/llvm-profdata/show-covered.test` so I need to re-generate its input, but I cannot find the generation logic. It was added in [11d3074](https://github.com/llvm/llvm-project/commit/11d3074267324e811ade248a991564fa6722754e)

Thanks for asking. I think you should not need to change this test, especially since this flag is off by default. I suspect it is broken because the variant mask bits were changed.

File: compiler-rt/include/profile/InstrProfData.inc
Comment (ronryvchin) [reply compiler-rt/include/profile/InstrProfData.inc:749]: Are bits 33-55 in use?

Comment (ronryvchin) [reply compiler-rt/include/profile/InstrProfData.inc]: I see that you are right (historically it started at `56` and climbed higher), but I cannot find usages of bit `55`

Comment (Dinistro) [reply compiler-rt/include/profile/InstrProfData.inc]: Considering that there is no usage of bit `55` and that there is no documentation about this being reserved, I would argue that we should be free to use `55`. If any downstream user is relying on such an undocumented implementation detail, we should be free to break it.

File: llvm/test/Transforms/PGOProfile/loop_entries_gen.ll
Comment (ellishg) [inline llvm/test/Transforms/PGOProfile/loop_entries_gen.ll:3]:
```suggestion
; RUN: opt %s -passes=pgo-instr-gen -S | FileCheck %s --check-prefixes=CHECK,NOTLOOPENTRIES --implicit-check-not=@llvm.instrprof.increment
; RUN: opt %s -passes=pgo-instr-gen -pgo-instrument-loop-entries -S | FileCheck %s --check-prefixes=CHECK,LOOPENTRIES --implicit-check-not=@llvm.instrprof.increment
; RUN: opt %s -passes=pgo-instr-gen -pgo-instrument-entry -S | FileCheck %s --check-prefixes=CHECK,FUNCTIONENTRY --implicit-check-not=@llvm.instrprof.increment
```

File: llvm/test/Transforms/PGOProfile/loop_entries_use.ll
Comment (ellishg) [inline llvm/test/Transforms/PGOProfile/loop_entries_use.ll:8]: You do not need to specify the flags when consuming profile because that info in encoded in the header.

```suggestion
; RUN: opt %t/main.ll -passes=pgo-instr-use -pgo-test-profile-file=%t/default.profdata -S | FileCheck %s
; RUN: llvm-profdata merge %t/loop_entries.proftext -o %t/loop_entries.profdata
; RUN: opt %t/main.ll -passes=pgo-instr-use -pgo-test-profile-file=%t/loop_entries.profdata -S | FileCheck %s
; RUN: llvm-profdata merge %t/function_entry.proftext -o %t/function_entry.profdata
; RUN: opt %t/main.ll -passes=pgo-instr-use -pgo-test-profile-file=%t/function_entry.profdata -S | FileCheck %s
```

**Comment by ronryvchin** [comment]
@ellishg Thank you for your comments!
I think everything is addressed now, is there anything else that should be done before merging?

**Review by ellishg — APPROVED** [review]
LGTM

**Pull request merged**
